### PR TITLE
Complete Refactor of Beacon Node With Single State

### LINKED ISSUE #781: Beacon Node Refactorization
With the finished [replacement](https://github.com/ethereum/eth2.0-specs/pull/122) of active and crystallized state with single state, we can begin to refactor beacon node to adopt these changes. The updates will touch base across multiple packages including core, DB, casper, types and will open up many design questions. Let's use this as a mega issue to track the overall progress, discuss the design questions, and document the design decisions. Everyone please provide your inputs here

### PR DESCRIPTION
**Pull request by rauljordan**
This resolves #781

---

# Description

This PR completes the refactor of the beacon node's core functionality into its own core package and eliminates usage of two states (crystallized/active) across the repo in favor of a single beacon state object.

## Changes

Two very big packages were moved / refactored: `casper` and `types`. In particular, we had a lot of important stateful logic spread across both of them in ways that made refactoring difficult and the core logic difficult to reason about. In the old code, for example, the `types/block.go` file contained validation logic that depended on the DB, important attestation checks that should instead be pure functions. Additionally, `types/crystallized_state.go` contained all the logic relevant to state transitions spread across a ton of functions in a file that should serve for struct definitions and basic methods. We instead restructured the logic into a single `core/` package containing stateless, pure functions that implement the logic defined in the ETH 2.0 spec.

Additionally, we refactored the rest of the services in the repo to only use a single `BeaconState` object where 2 were used before. The DB schema and basic protos have also been modified to remove references to having 2 states and a `BeaconState` proto definition was created with the fields defined in the spec.

## Important Files to Look At

The `core` package contains 4 important subpackages:
- `types`
- `state`
- `incentives`
- `validators`

Types serves to define Go structs around primitives of Serenity such as blocks, the state, attestations, and others. Note that we aim to keep these files simple, as they are mostly getters/setters for underlying proto representations along with some logic to verify or transform data as needed by these structs. We _avoid_ placing important, critical logic into this package.

The `incentives` package serves to define all reward/penalty conditions for Casper FFG as used by the beacon node. It defines pure functions for calculating incentives as well as functions that calculate items related to attestation votes.

The `validators` package defines all pure functions that perform transformations on validators of the beacon chain. We include functions that help us rotate validators (enter/exit), penalize validators, shuffle validators, determine attester/proposer indices in the set, and calculate shard committees. The logic in this package is all pure functions.

The `state` package defines validity checks for primitives of the system based on the spec as well as the state transition function to be used by Serenity. Once again, this package is stateless and does not assume anything about services which will be using it. It just defines a pure function that applies the steps defined in the state transition on a per-block and per-epoch basis in the ETH 2.0 spec.

Additionally, this includes helper functions that make the state transition more readable and easier to unit test in smaller components such as processing cross links and special records.

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -8,9 +8,11 @@ import (
 	"fmt"
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: Please review this file - this changes the blockchain service to use a single state

File: beacon-chain/core/state/state_transition.go [hunk 2]
```diff
@@ -1,6 +1,9 @@
 package state
```
Comment (rauljordan) [inline beacon-chain/core/state/state_transition.go]: This file basically merges the computation of a new active state with a new crystallized state into a single function

File: beacon-chain/core/state/validity_conditions.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package state
```
Comment (rauljordan) [inline beacon-chain/core/state/validity_conditions.go]: Validity conditions will be moved here in order to prevent import cycles and keep types/block.go clean

File: beacon-chain/core/types/attestation.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
+package types
```
Comment (rauljordan) [inline beacon-chain/core/types/attestation.go]: Copy paste, no need to review

File: beacon-chain/core/types/attestation_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package types
```
Comment (rauljordan) [inline beacon-chain/core/types/attestation_test.go]: Copy paste, no need to review

File: beacon-chain/core/types/block.go [hunk 6]
```diff
@@ -1,9 +1,20 @@
 package types
```
Comment (rauljordan) [inline beacon-chain/core/types/block.go]: Copy paste, no need to review

File: beacon-chain/core/types/state.go [hunk 7]
```diff
@@ -1,6 +1,8 @@
 package types
```
Comment (rauljordan) [inline beacon-chain/core/types/state.go]: Merged methods and fields from both active/crystallized into one

File: beacon-chain/db/block_test.go [hunk 8]
```diff
@@ -3,7 +3,7 @@ package db
 import (
```
Comment (rauljordan) [inline beacon-chain/db/block_test.go]: Copy paste, no need to review

File: beacon-chain/db/state.go [hunk 9]
```diff
@@ -5,34 +5,26 @@ import (
```
Comment (rauljordan) [inline beacon-chain/db/state.go]: Modified db methods to only use a single state

File: beacon-chain/rpc/service.go [hunk 10]
```diff
@@ -11,9 +11,9 @@ import (
```
Comment (rauljordan) [inline beacon-chain/rpc/service.go]: Modified RPC to use a single state

File: beacon-chain/rpc/service_test.go [hunk 11]
```diff
@@ -11,8 +11,8 @@ import (
 	"github.com/golang/mock/gomock"
```
Comment (rauljordan) [inline beacon-chain/rpc/service_test.go]: Copy paste, no need to review

File: beacon-chain/simulator/service.go [hunk 12]
```diff
@@ -5,12 +5,13 @@ import (
 	"bytes"
```
Comment (rauljordan) [inline beacon-chain/simulator/service.go]: Modified simulator to use single beacon state

File: beacon-chain/simulator/service_test.go [hunk 13]
```diff
@@ -58,7 +58,7 @@ func setupSimulator(t *testing.T, beaconDB *db.BeaconDB) (*Simulator, *mockP2P)
 		Web3Service:     &mockPOWChainService{},
```
Comment (rauljordan) [inline beacon-chain/simulator/service_test.go]: Copy paste, no need to review

File: beacon-chain/blockchain/service.go [hunk 14]
```diff
@@ -195,43 +189,26 @@ func (c *ChainService) updateHead(processedBlock <-chan *types.Block) {
 			}
 
 			// TODO(#674): Handle chain reorgs.
-			var newCState *types.CrystallizedState
-			if c.unfinalizedBlocks[h].cycleTransition {
-				newCState = blockcState
+			var newState *types.BeaconState
+			if c.unfinalizedBlocks[h].IsCycleTransition(block.SlotNumber()) {
```
Comment (rawfalafel) [inline beacon-chain/blockchain/service.go]: Now that the state's updated every slot, this if statement shouldn't be necessary.

File: beacon-chain/blockchain/service.go [hunk 15]
```diff
@@ -195,43 +189,26 @@ func (c *ChainService) updateHead(processedBlock <-chan *types.Block) {
...
 				continue
 			}
 			log.WithField("blockHash", fmt.Sprintf("0x%x", h)).Info("Chain head block and state updated")
 			// We fire events that notify listeners of a new block (or crystallized state in
 			// the case of a state transition). This is useful for the beacon node's gRPC
 			// server to stream these events to beacon clients.
-			if c.unfinalizedBlocks[h].cycleTransition {
-				c.canonicalCrystallizedStateFeed.Send(blockcState)
+			if c.unfinalizedBlocks[h].IsCycleTransition(block.SlotNumber()) {
+				c.canonicalStateFeed.Send(newState)
```
Comment (rawfalafel) [inline beacon-chain/blockchain/service.go]: Merging `active` and `crystallized` causes the same issue here. If I recall correctly, subscribers for this feed care about new committee assignments. If so, we may want to reconsider some of the logic here to make the intention more clear.

File: beacon-chain/blockchain/service.go [hunk 16]
```diff
@@ -275,80 +252,87 @@ func (c *ChainService) processBlock(block *types.Block) error {
...
 	// so that they're accounted for during cycle transitions.
-	aState = aState.UpdateAttestations(block.Attestations())
+	beaconState.SetPendingAttestations(block.Attestations())
 
 	// If the block is valid, we compute its associated state tuple (active, crystallized)
-	var didCycleTransition bool
-	if cState.IsCycleTransition(block.SlotNumber()) {
-		cState, err = c.executeStateTransition(cState, aState, block)
+	if beaconState.IsCycleTransition(block.SlotNumber()) {
+		beaconState, err = c.executeStateTransition(beaconState, block, parent.SlotNumber())
```
Comment (rawfalafel) [inline beacon-chain/blockchain/service.go]: I find it fishy that this state transition function is calculating the next state by accepting the block of that next state as a parameter. This may be partially contributing to the circular reference problem between blocks and states. This sorta works in this situation since the next block is given to us, but will be an awkward interface when a local validator wants to generate the next block.

File: beacon-chain/db/state.go [hunk 17]
```diff
@@ -44,169 +36,98 @@ func (db *BeaconDB) InitializeState(genesisValidators []*pb.ValidatorRecord) err
...
-	})
-
-	return aState, err
-}
-
-// GetCrystallizedState fetches the current active state.
-func (db *BeaconDB) GetCrystallizedState() (*types.CrystallizedState, error) {
-	var cState *types.CrystallizedState
+// GetState --
+func (db *BeaconDB) GetState() (*types.BeaconState, error) {
```
Comment (rawfalafel) [inline beacon-chain/db/state.go]: We want to eventually have this accept a hash as a parameter, so that previous states can be retrieved as well, correct?

File: beacon-chain/blockchain/service.go (hunk 15 above)
Comment (rauljordan) [reply beacon-chain/blockchain/service.go]: you mean they are about knowing when the validator set changed?

File: beacon-chain/core/state/state_transition_test.go [hunk 18]
```diff
@@ -49,12 +48,12 @@ func TestInitialDeriveCrystallizedState(t *testing.T) {
...
-		t.Fatalf("expected last state recalc to equal %d: got %d", params.BeaconConfig().CycleLength, newState.LastStateRecalculationSlot())
+	if newState.LastStateRecalculationSlot() != 1 {
+		t.Fatalf("expected last state recalc to equal %d: got %d", 1, newState.LastStateRecalculationSlot())
 	}
 
 	if newState.LastFinalizedSlot() != 0 {
-		t.Fatalf("xpected finalized slot to equal %d, got %d", 0, newState.LastFinalizedSlot())
+		t.Fatalf("expected finalized slot to equal %d, got %d", 0, newState.LastFinalizedSlot())
 	}
 }
```
Comment (rauljordan) [inline beacon-chain/core/state/state_transition_test.go]: Please review this test @terenc3t @rawfalafel

File: beacon-chain/db/block.go [hunk 19]
```diff
@@ -140,28 +133,21 @@ func (db *BeaconDB) UpdateChainHead(block *types.Block, aState *types.ActiveStat
...
 		}
 
-		if err := chainInfo.Put(aStateLookupKey, aStateEnc); err != nil {
-			return fmt.Errorf("failed to save active state as canonical: %v", err)
-		}
-
-		if cStateEnc != nil {
-			if err := chainInfo.Put(cStateLookupKey, cStateEnc); err != nil {
-				return fmt.Errorf("failed to save crystallized state as canonical: %v", err)
-			}
```
Comment (nisdas) [inline beacon-chain/db/block.go]: it should have the conditional first
```go
if beaconStateEnc != nil {
```

File: beacon-chain/db/state.go [hunk 20]
```diff
@@ -5,34 +5,26 @@ import (
...
-	aState := types.NewGenesisActiveState()
-	cState, err := types.NewGenesisCrystallizedState(genesisValidators)
+	beaconState, err := types.NewGenesisBeaconState(genesisValidators)
 	if err != nil {
 		return err
 	}
 
 	// #nosec G104
-	activeStateHash, _ := aState.Hash()
+	stateHash, _ := beaconState.Hash()
```
Comment (nisdas) [inline beacon-chain/db/state.go]: Any reason, we are throwing away all the errors here ?

File: beacon-chain/db/state.go (hunk 17 above)
Comment (nisdas) [reply beacon-chain/db/state.go]: Wouldnt this be just referring to the current state ? If we wanted to retrieve other previous states, we would need it to be in a separate bucket.

File: beacon-chain/simulator/service.go [hunk 21]
```diff
@@ -51,29 +52,19 @@ type Config struct {
...
-	CStateReqBuf        int
+	StateReqBuf         int
 	P2P                 p2pAPI
 	Web3Service         powChainService
-	BeaconDB            beaconDB
+	BeaconDB            *db.BeaconDB
 	EnablePOWChain      bool
 }
 
-type beaconDB interface {
```
Comment (nisdas) [inline beacon-chain/simulator/service.go]: Why was this removed ? We would only want to expose this service to methods that it will actually use.

File: beacon-chain/core/state/state_transition.go [hunk 22]
```diff
@@ -26,81 +30,96 @@ func NewStateTransition(
...
-	newState.SetCrossLinks(crossLinks)
+	newRandao := createRandaoMix(block.RandaoReveal(), st.RandaoMix())
+	newState.SetRandaoMix(newRandao[:])
 
-	newState.SetLastJustifiedSlot(justifiedSlot)
-	newState.SetLastFinalizedSlot(finalizedSlot)
-	newState.SetJustifiedStreak(justifiedStreak)
-	newState.SetLastStateRecalculationSlot(newState.LastStateRecalculationSlot() + params.BeaconConfig().CycleLength)
+	// The changes below are only applied if this is a cycle transition.
+	if block.SlotNumber()%params.BeaconConfig().CycleLength == 0 {
```
Comment (rawfalafel) [inline beacon-chain/core/state/state_transition.go]: Why did this logic change from before? A cycle transition can also occur if `block.slot % cyleLength == 1` if the previous slot was skipped, and this needs to also handle the corner case where multiple cycles worth of slots are skipped.

File: beacon-chain/blockchain/service.go [hunk 23]
```diff
@@ -275,80 +252,88 @@ func (c *ChainService) processBlock(block *types.Block) error {
...
-
-	aState, err = aState.CalculateNewActiveState(
-		block,
-		cState,
-		parent.SlotNumber(),
-	)
+	beaconState, err = c.executeStateTransition(beaconState, block, parent.SlotNumber())
 	if err != nil {
-		return fmt.Errorf("compute active state failed: %v", err)
+		return fmt.Errorf("initialize new cycle transition failed: %v", err)
```
Comment (rawfalafel) [inline beacon-chain/blockchain/service.go]: [np] execute state transition failed

**Comment by rauljordan** [comment]
Closing to reopen in a new target branch
